staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go

```diff
@@ -501,3 +505,166 @@ func TestReflectorForWatchCache(t *testing.T) {
+			eventCount:         5,
+			cacheCapacity:      5,
+			lowerBoundCapacity: 3,
+			upperBoundCapacity: 5 * 2,
+			interval:           eventFreshDuration/2 + time.Second,
+			expectCapacity:     3,
+		},
+		{
+			name:               "[singular capacity] events inside eventFreshDuration cause cache expanding with given upperBoundCapacity",
+			eventCount:         5,
+			cacheCapacity:      5,
+			lowerBoundCapacity: 5 / 2,
+			upperBoundCapacity: 8,
+			interval:           eventFreshDuration / 6,
+			expectCapacity:     8,
+		},
+		{
+			name:               "[even capacity] events inside eventFreshDuration cause cache expanding",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 7,
+			expectCapacity:     12,
+		},
+		{
+			name:               "[even capacity] events outside eventFreshDuration without change cache capacity",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 6,
+			expectCapacity:     6,
+		},
+		{
+			name:               "[even capacity] half recent events outside eventFreshDuration cause cache shrinking",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 2,
+			expectCapacity:     3,
+		},
+		{
+			name:               "[even capacity] half recent events outside eventFreshDuration cause cache shrinking with given lowerBoundCapacity",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 4,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 2,
+			expectCapacity:     4,
+		},
+		{
+			name:               "[even capacity] events inside eventFreshDuration cause cache expanding with given upperBoundCapacity",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 10,
+			interval:           eventFreshDuration / 7,
+			expectCapacity:     10,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			store := newTestWatchCache(test.cacheCapacity, &cache.Indexers{})
+			store.cache = store.cache[:0]
+			store.lowerBoundCapacity = test.lowerBoundCapacity
+			store.upperBoundCapacity = test.upperBoundCapacity
+			loadEventWithDuration(store, test.eventCount, test.interval)
+			nextInterval := store.clock.Now().Add(time.Duration(test.interval.Nanoseconds() * int64(test.eventCount)))
+			store.resizeCache(nextInterval)
+			if store.capacity != test.expectCapacity {
+				t.Errorf("expect capacity %d, but get %d", test.expectCapacity, store.capacity)
+			}
+			// check last event's index after cache resizing.
+			key := fmt.Sprintf("event-%d", test.eventCount-1)
```

Comment:
Instead of this logic here, I would prefer checking the following
- checking if startIndex and endIndex are set correctly
- going over all elements from startIndex to endIndex and ensuring that they have correct keys

Reply:
Done.